### PR TITLE
cmd/shard: new command

### PR DESCRIPTION
**Pull request by razvanazamfirei**
- [X] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [X] Have you successfully run `brew style` with your changes locally?
- [X] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This PR adds `brew shard`, a command used to shard repositories and, more importantly, audit that files are in the correct place. 

Still needs tests written and making sure flags work as intended, but I appreciate any feedback.

Currently I'm using `git mv` for moving files in order to preserve history. It's a bit slow, but this should be noticeable only on the initial sharding.

### CODE REVIEW

File: Library/Homebrew/cmd/shard.rb
Comment (carlocab) [inline Library/Homebrew/cmd/shard.rb:1]: Probably should be in `dev-cmd`?

File: Library/Homebrew/cmd/shard.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        directory is specified as a command-line argument.
+      EOS
+      flag "--directory=",
+           description: "Specify the starting directory"
+      flag "--file_limit=",
+           description: "Specify the maximum number of files per-directory before sharding."
+      flag "--file_subset=",
+           description: "Limit to a subset of files based on initial letter."
+      switch "--check",
+             description: "Audit the directory for sharding."
```
Comment (carlocab) [inline Library/Homebrew/cmd/shard.rb:22]: Should this just be done by default?

File: Library/Homebrew/sharder.rb [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    end
+
+    subfolders.each do |subfolder|
+      Sharder.new(directory: subfolder, file_limit: file_limit, file_subset:
+      file_subset).shard
+    end
+  end
+
+  def move_file(source, destination)
+    success = system("git mv '#{source}' '#{destination}' 2>/dev/null")
```
Comment (carlocab) [inline Library/Homebrew/sharder.rb:44]: You should probably use `system_command` here

File: Library/Homebrew/sharder.rb [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    FileUtils.mv(source, destination) unless success
+  rescue Errno::ENOENT
+    raise "Error moving file from #{source} to #{destination}"
```
Comment (carlocab) [inline Library/Homebrew/sharder.rb:47]:
```suggestion
    onoe "Error moving file from #{source} to #{destination}"
    raise
```

File: Library/Homebrew/sharder.rb [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  end
+
+  def shard_subfolder(subfolder)
+    if (subfolder.children.size - 2) > file_limit
+      self.class.new(directory: subfolder.to_s, file_limit: file_limit, file_subset: file_subset)
+          .shard
+      FileUtils.rm_rf(subfolder)
+    end
+  rescue Errno::ENOENT
+    raise "Error removing directory: #{subfolder}"
```
Comment (carlocab) [inline Library/Homebrew/sharder.rb:57]:
```suggestion
    onoe "Error removing directory: #{subfolder}"
    raise
```

File: Library/Homebrew/sharder.rb [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+require "pathname"
+
+class Sharder
+  DEFAULT_DIRECTORY = Pathname.new("Casks").freeze
+  DEFAULT_MAX_FILES = 100
+  FILE_SUBSET = nil
+
+  attr_reader :directory, :file_limit, :file_subset
+
+  def initialize(directory: DEFAULT_DIRECTORY, file_limit: DEFAULT_MAX_FILES, file_subset: FILE_SUBSET)
```
Comment (carlocab) [inline Library/Homebrew/sharder.rb:15]: Might be nice to add type annotations to methods here.

File: Library/Homebrew/sharder.rb [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+  def move_file(source, destination)
+    success = system("git mv '#{source}' '#{destination}' 2>/dev/null")
+    FileUtils.mv(source, destination) unless success
+  rescue Errno::ENOENT
+    raise "Error moving file from #{source} to #{destination}"
+  end
+
+  def shard_subfolder(subfolder)
+    if (subfolder.children.size - 2) > file_limit
```
Comment (carlocab) [inline Library/Homebrew/sharder.rb:51]: The `- 2` is a bit cryptic here.

File: Library/Homebrew/sharder.rb [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+      self.class.new(directory: subfolder.to_s, file_limit: file_limit, file_subset: file_subset)
+          .shard
+      FileUtils.rm_rf(subfolder)
```
Comment (carlocab) [inline Library/Homebrew/sharder.rb:54]:
```suggestion
      subfolder.rmtree
```
Maybe?

File: Library/Homebrew/sharder.rb [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+      lost_files << file if file.dirname.to_s != subdir.to_s
+    end
+    lost_files
+  end
+
+  def relocate_files(lost_files)
+    lost_files.each do |file|
+      subdir = destination_dir(file.basename.to_s, true)
+      FileUtils.mkdir_p(directory.join(subdir))
+      move_file(file, directory.join(subdir, file.basename))
```
Comment (carlocab) [inline Library/Homebrew/sharder.rb:76]:
```suggestion
      (directory/subdir).mkpath
      move_file(file, directory/subdir/file.basename)
```

File: Library/Homebrew/sharder.rb [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    raise "Error removing directory: #{subfolder}"
+  end
+
+  def check
+    lost_files = []
+    Find.find(directory.to_s) do |path|
+      file = Pathname.new(path)
+      next unless file.file?
+
+      subdir = directory.join(destination_dir(file.basename.to_s, true))
```
Comment (carlocab) [inline Library/Homebrew/sharder.rb:66]:
```suggestion
      subdir = directory/destination_dir(file.basename.to_s, true)
```

File: Library/Homebrew/sharder.rb [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+      lost_files << file if file.dirname.to_s != subdir.to_s
```
Comment (carlocab) [inline Library/Homebrew/sharder.rb:67]:
```suggestion
      lost_files << file if file.dirname.realpath != subdir.realpath
```

File: Library/Homebrew/sharder.rb [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    subfolders = directory.children.select(&:directory?)
+
+    should_distribute = files.size > file_limit
+
+    files.each do |file|
+      next if file_subset&.exclude?(file.basename.to_s[0].downcase)
+
+      subdir = destination_dir(file.basename.to_s, should_distribute)
+      FileUtils.mkdir_p(directory.join(subdir))
+      move_file(file, directory.join(subdir, file.basename))
```
Comment (carlocab) [inline Library/Homebrew/sharder.rb:34]:
```suggestion
      (directory/subdir).mkpath
      move_file(file, directory/subdir/file.basename)
```

**Review by carlocab — COMMENTED** [review]
Could you describe your intended sharding scheme/strategy here more explicitly?

**Comment by razvanazamfirei** [comment]
@carlocab, at least for `homebrew-casks` the [scheme](https://github.com/Homebrew/homebrew-cask/pull/141893) would be: 
```
#{first-letter}/#{first-letter}#{second-letter}
```
For sharding purposes, all non-alphabetical characters are replaced by `_`. This does create a case where `0-ad.rb` ends up in `_/__/0-ad.rb`. If there's a better character, happy to change it.

Not a fan of using
```
"git mv '#{source}' '#{destination}' 2>/dev/null"
```
The rationale behind it was to try and preserve the file history. `system_command` is the fallback (and much faster). 

I'll integrate the rest of the comments.

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]
Thanks for the PR @razvanazamfirei, it's much appreciated and I love the forward progress on sharding but I don't think it's quite the right approach for us for a few reasons:
- I don't think we need either a user-focused or developer-focused command for something that will be run only once per tap
- the auditing process should be moved into `brew audit` (or `brew style` properly but we should mostly do the right thing if they run `brew create` for something that's going to end up in the right tap
- similarly, there may be a need to teach other Homebrew internals how to generate the right shard name and this sharding logic will likely vary per-tap and is only needed on two taps (homebrew-core and homebrew-cask)

I'll reply to a couple of other things in the issue thread too.


Thanks again ❤️

**Comment by carlocab** [comment]
Just leaving a note here: `brew formulae` seems to be broken by sharding. If you have `homebrew/test-bot` tapped, you see `homebrew/test-bot/t/testbottest` in the output.

**Pull request closed without merging**